### PR TITLE
[Enhance] [QueryEngine] add OlapScanPrepareOperator

### PR DESCRIPTION
**Pull request by ZiheLiu**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Problem Summary：
Concurrency scanning on the single tablet need the information, including **conjuncts** and **key ranges** parsed from conjuncts.
Normalizing conjuncts and parsing key ranges can only happen, after the local runtime filters have been built.

Therefore, dividing the tablet to concurrency scan can only happen, after the local runtime filters built and before `OlapScanOperator` starts to pull_chunk.
Add a new `OlapScanPrepareOperator` to implement this.

## Modifaction
- Add a new `OlapScanPrepareOperator`
    - `DOP` is 1.
    - Normalizing conjuncts and parsing key ranges at the first and only time calling `OlapScanPrepareOperator::pull_chunk`, and remove this logic in each `OlapChunkSource`.
- Add a new `NoopSinkOperator`
    - Because a driver must contain at least 2 operators, we need a placeholder sink operator to make `OlapScanPrepareOperator` become a driver.
- `ScanOperator`
    - Remove prepare/close `_scan_node->conjunct_ctxs` in `ScanOperator::prepare/close`, and add it to `OlapScanPrepareOperator` and `ConnectorScanOperator`.
    - Remove init runtime filter in `decompose_scan_node_to_pipeline`, and add it to `OlapScanNode::decompose_to_pipeline` and `ConnectorScanNode::decompose_to_pipeline`.
- `OlapScanOperator`
    - Move `DictOptimizeParser::rewrite_descriptor` logic from ``OlapScanOperatorFactory::prepare`` to `OlapScanPrepareOperatorFactory::prepare`
    - `OlapScanOperator` should wait `OlapScanPrepareOperator` finished.

### CODE REVIEW

File: be/src/exec/pipeline/scan/olap_scan_context.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    const TOlapScanNode& thrift_olap_scan_node = _scan_node->thrift_olap_scan_node();
+    const TupleDescriptor* tuple_desc = state->desc_tbl().get_tuple_descriptor(thrift_olap_scan_node.tuple_id);
+
+    // Get _conjunct_ctxs.
+    _conjunct_ctxs = _scan_node->conjunct_ctxs();
+    _conjunct_ctxs.insert(_conjunct_ctxs.end(), runtime_in_filters.begin(), runtime_in_filters.end());
+
+    // eval_const_conjuncts.
+    Status status;
+    RETURN_IF_ERROR(vectorized::OlapScanConjunctsManager::eval_const_conjuncts(_conjunct_ctxs, &status));
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/olap_scan_context.cpp:36]: Looks like this function could return `Status` directly ?

File: be/src/exec/pipeline/scan/scan_operator.h [hunk 2]
```diff
@@ -36,6 +36,8 @@ class ScanOperator : public SourceOperator {
     void set_workgroup(workgroup::WorkGroupPtr wg);
 
     // interface for different scan node
+    virtual bool maybe_has_output() const { return true; }
+    virtual bool must_be_finished() const { return false; }
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/scan_operator.h:40]: Leave some comment? What does `must_be_finished` means?

File: be/src/exec/pipeline/scan/olap_chunk_source.cpp [hunk 3]
```diff
@@ -126,27 +99,6 @@ void OlapChunkSource::_init_counter(RuntimeState* state) {
     _io_timer = ADD_TIMER(_runtime_profile, "IOTime");
 }
 
-Status OlapChunkSource::_build_scan_range(RuntimeState* state) {
-    RETURN_IF_ERROR(_conjuncts_manager.get_key_ranges(&_key_ranges));
-    _conjuncts_manager.get_not_push_down_conjuncts(&_not_push_down_conjuncts);
-    _dict_optimize_parser.rewrite_conjuncts<false>(&_not_push_down_conjuncts, state);
-
-    // FixMe(kks): Ensure this logic is right.
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/olap_chunk_source.cpp:134]: @kangkaisen Could you ensure it or remove the comment?

File: be/src/exec/pipeline/scan/olap_scan_context.cpp (hunk 1 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/olap_scan_context.cpp:36]: It's a little misunderstanding. However, Non-pipeline uses it as the follow codes. The return status `eval_const_conjuncts` indicates there is something wrong at evaluating, while the augment status is is guaranteed to be `OK` or `EndOfFile`.
```C++
Status OlapScanNode::open(RuntimeState* state) {
    SCOPED_TIMER(_runtime_profile->total_time_counter());
    RETURN_IF_CANCELLED(state);
    RETURN_IF_ERROR(ExecNode::open(state));

    Status status;
    RETURN_IF_ERROR(OlapScanConjunctsManager::eval_const_conjuncts(_conjunct_ctxs, &status));
    _update_status(status);

    _dict_optimize_parser.set_mutable_dict_maps(state, state->mutable_query_global_dict_map());
    DictOptimizeParser::rewrite_descriptor(state, _conjunct_ctxs, _olap_scan_node.dict_string_id_to_int_ids,
                                           &(_tuple_desc->decoded_slots()));

    return Status::OK();
}
```

File: be/src/exec/pipeline/scan/scan_operator.h (hunk 2 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/scan_operator.h]: It is to change the behaviour of `has_output` and `is_finished` by the child `OlapScanOpeartor`. 
However, it is hard to understand, so I remove them, and make `OlapScanOperator` overrides `has_output` and `is_finished`.

File: be/src/exec/pipeline/scan/olap_chunk_source.cpp (hunk 3 above)
Comment (kangkaisen) [reply be/src/exec/pipeline/scan/olap_chunk_source.cpp:134]: The logic should be right and the comment could be removed.

File: be/src/exec/vectorized/olap_scan_node.cpp [hunk 4]
```diff
@@ -642,8 +644,27 @@ void OlapScanNode::_close_pending_scanners() {
 }
 
 pipeline::OpFactories OlapScanNode::decompose_to_pipeline(pipeline::PipelineBuilderContext* context) {
-    auto factory = std::make_shared<pipeline::OlapScanOperatorFactory>(context->next_operator_id(), this);
-    return pipeline::decompose_scan_node_to_pipeline(factory, this, context);
+    auto scan_ctx = std::make_shared<pipeline::OlapScanContext>(this);
+    auto&& rc_rf_probe_collector = std::make_shared<RcRfProbeCollector>(2, std::move(this->runtime_filter_collector()));
+
+    // scan_prepare_op.
+    auto scan_prepare_op =
```
Comment (kangkaisen) [inline be/src/exec/vectorized/olap_scan_node.cpp:651]: Is there will be performance issue if add additional two operators?

Comment (ZiheLiu) [reply be/src/exec/vectorized/olap_scan_node.cpp:651]: I test on SSB single dataset, and there isn't performance regression.
The only one additional work is schedule of this two operators.

**Review by kangkaisen — DISMISSED** [review]

**Review by murphyatwork — DISMISSED** [review]

**Comment by kangkaisen** [comment]
run starrocks_be_unittest

**Comment by kangkaisen** [comment]
run starrocks_be_unittest

**Comment by ZiheLiu** [comment]
run starrocks_be_unittest

**Comment by ZiheLiu** [comment]
run starrocks_be_unittest

**Review by silverbullet233 — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**